doc/doxygen/man/man3/Jugada.3
.TH "Jugada" 3 "Dl Jun 1 2020" "Version v3" "Escacs" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Jugada \- Conte tota la informacio de una \fBJugada\fP i part de les dades de una partida\&.  

.SH SYNOPSIS
.br
.PP
.SS "Mètodes públics"

.in +1c
.ti -1c
.RI "\fBJugada\fP ()"
.br
.ti -1c
.RI "\fBJugada\fP (\fBPosicio\fP \fBorigen\fP, \fBPosicio\fP \fBdesti\fP, \fBColor\fP \fBtirador\fP)"
.br
.ti -1c
.RI "\fBJugada\fP (\fBPosicio\fP \fBorigen\fP, \fBPosicio\fP \fBdesti\fP, \fBPeca\fP \fBpecaOrigen\fP, \fBPeca\fP \fBpecaDesti\fP)"
.br
.ti -1c
.RI "\fBJugada\fP (\fBResultatJugada\fP resultat, \fBColor\fP \fBtirador\fP)"
.br
.ti -1c
.RI "void \fBafegirResultatJugada\fP (\fBResultatJugada\fP res)"
.br
.ti -1c
.RI "void \fBcanviTauler\fP (\fBTauler\fP \fBt\fP)"
.br
.ti -1c
.RI "void \fBfinalitzarJugada\fP (\fBTauler\fP \fBt\fP, boolean partidaAcabada, \fBResultatPartida\fP \fBresultatFinal\fP, \fBContadors\fP \fBcontador\fP, \fBPeca\fP \fBpecaPromocionada\fP, \fBColor\fP \fBguanyador\fP)"
.br
.ti -1c
.RI "boolean \fBobtenirAcabada\fP ()"
.br
.ti -1c
.RI "\fBContadors\fP \fBobtenirContadors\fP ()"
.br
.ti -1c
.RI "\fBPosicio\fP \fBobtenirDesti\fP ()"
.br
.ti -1c
.RI "\fBColor\fP \fBobtenirGuanyador\fP ()"
.br
.ti -1c
.RI "\fBPosicio\fP \fBobtenirOrigen\fP ()"
.br
.ti -1c
.RI "\fBPeca\fP \fBobtenirPecaDesti\fP ()"
.br
.ti -1c
.RI "\fBPeca\fP \fBobtenirPecaOrigen\fP ()"
.br
.ti -1c
.RI "\fBTipusPeca\fP \fBobtenirPecaPromocionada\fP ()"
.br
.ti -1c
.RI "\fBResultatPartida\fP \fBobtenirResultatPartida\fP ()"
.br
.ti -1c
.RI "Collection< \fBResultatJugada\fP > \fBobtenirResultatsJugada\fP ()"
.br
.ti -1c
.RI "\fBTauler\fP \fBobtenirTauler\fP ()"
.br
.ti -1c
.RI "\fBColor\fP \fBobtenirTirador\fP ()"
.br
.in -1c
.SS "Atributs Privats"

.in +1c
.ti -1c
.RI "boolean \fBacabada\fP"
.br
.ti -1c
.RI "\fBContadors\fP \fBcontador\fP"
.br
.ti -1c
.RI "\fBPosicio\fP \fBdesti\fP"
.br
.ti -1c
.RI "\fBColor\fP \fBguanyador\fP"
.br
.ti -1c
.RI "\fBPosicio\fP \fBorigen\fP"
.br
.ti -1c
.RI "\fBPeca\fP \fBpecaDesti\fP"
.br
.ti -1c
.RI "\fBPeca\fP \fBpecaOrigen\fP"
.br
.ti -1c
.RI "\fBPeca\fP \fBpecaPromocionada\fP"
.br
.ti -1c
.RI "\fBResultatPartida\fP \fBresultatFinal\fP"
.br
.ti -1c
.RI "Queue< \fBResultatJugada\fP > \fBresultatsJugada\fP"
.br
.ti -1c
.RI "\fBTauler\fP \fBt\fP"
.br
.ti -1c
.RI "\fBColor\fP \fBtirador\fP"
.br
.in -1c
.SH "Descripció Detallada"
.PP 
Conte tota la informacio de una \fBJugada\fP i part de les dades de una partida\&. 


.PP
\fBAutor\fP
.RS 4
Eloi Quintana Ferrer 
.PP
u1963141 
.RE
.PP

.SH "Documentació del Constructor i el Destructor"
.PP 
.SS "Jugada\&.Jugada ()"

.PP
\fBPrecondició\fP
.RS 4
@Post constructor 
.RE
.PP

.SS "Jugada\&.Jugada (\fBPosicio\fP origen, \fBPosicio\fP desti, \fBColor\fP tirador)"

.PP
\fBPrecondició\fP
.RS 4
origen != null, desti != null, tirador != null, 
.PP
origen i desti dins el tauler @Post S'han guardat les variables d'entrada\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIorigen\fP Posició d'origen de la peça\&. 
.br
\fIdesti\fP Posició de desti de la peça\&. 
.br
\fItirador\fP Tirador de la jugada\&. 
.RE
.PP

.SS "Jugada\&.Jugada (\fBPosicio\fP origen, \fBPosicio\fP desti, \fBPeca\fP pecaOrigen, \fBPeca\fP pecaDesti)"

.PP
\fBPrecondició\fP
.RS 4
origen != null, desti != null, pecaOrigen != null, 
.PP
origen i desti dins el tauler @Post S'han guardat les variables d'entrada\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIorigen\fP Posició d'origen de la peça\&. 
.br
\fIdesti\fP Posició de desti de la peça\&. 
.br
\fIpecaOrigen\fP Peça de l'origen\&. 
.br
\fIpecaDesti\fP Peça del desti (posible peça del rival elimanada)\&. 
.br
\fIresultat\fP Resultat de la jugada\&. 
.br
\fIeliminades\fP Peces que s'han eliminat\&. 
.RE
.PP

.SS "Jugada\&.Jugada (\fBResultatJugada\fP resultat, \fBColor\fP tirador)"

.PP
\fBPrecondició\fP
.RS 4
resultat != null @Post S'han guardat les variables d'entrada\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIresultat\fP Resultat de la \fBJugada\fP\&. 
.br
\fItirador\fP color del jugador que realitza la jugada 
.RE
.PP

.SH "Documentació de les Funcions Membre"
.PP 
.SS "void Jugada\&.afegirResultatJugada (\fBResultatJugada\fP res)"

.PP
\fBPrecondició\fP
.RS 4
res != null @Post s'ha afegit a la llista el resultat de la jugada 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIt\fP \fBTauler\fP referenciat\&. 
.RE
.PP

.SS "void Jugada\&.canviTauler (\fBTauler\fP t)"

.PP
\fBPrecondició\fP
.RS 4
t != null @Post S'ha substituit la referencia de tauler amb la d'entrada 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIt\fP \fBTauler\fP referenciat\&. 
.RE
.PP

.SS "void Jugada\&.finalitzarJugada (\fBTauler\fP t, boolean partidaAcabada, \fBResultatPartida\fP resultatFinal, \fBContadors\fP contador, \fBPeca\fP pecaPromocionada, \fBColor\fP guanyador)"

.PP
\fBPrecondició\fP
.RS 4
- 
.RE
.PP
\fBPostcondició\fP
.RS 4
S'han assignat els parametrs d'entrada a les corresponents varialbes/referencies\&. 
.RE
.PP
\fBParàmetres\fP
.RS 4
\fIt\fP \fBTauler\fP referenciat\&. 
.br
\fIpartidaAcabada\fP true si la partida es acabada, false altrament\&. 
.br
\fIresultatFinal\fP resultat de la partida 
.br
\fIcontador\fP contadors de innacció i escac continu\&. 
.RE
.PP

.SS "boolean Jugada\&.obtenirAcabada ()"

.PP
\fBRetorna\fP
.RS 4
True si la partida esta acabada, false sino\&. 
.RE
.PP

.SS "\fBContadors\fP Jugada\&.obtenirContadors ()"

.PP
\fBRetorna\fP
.RS 4
\fBContadors\fP de la partida\&. 
.RE
.PP

.SS "\fBPosicio\fP Jugada\&.obtenirDesti ()"

.PP
\fBRetorna\fP
.RS 4
\fBPosicio\fP de desti 
.RE
.PP

.SS "\fBColor\fP Jugada\&.obtenirGuanyador ()"

.PP
\fBRetorna\fP
.RS 4
retorna el color del guanyador 
.RE
.PP

.SS "\fBPosicio\fP Jugada\&.obtenirOrigen ()"

.PP
\fBRetorna\fP
.RS 4
\fBPosicio\fP d'origen 
.RE
.PP

.SS "\fBPeca\fP Jugada\&.obtenirPecaDesti ()"

.PP
\fBRetorna\fP
.RS 4
peca de desti 
.RE
.PP

.SS "\fBPeca\fP Jugada\&.obtenirPecaOrigen ()"

.PP
\fBRetorna\fP
.RS 4
peca d'origen 
.RE
.PP

.SS "\fBTipusPeca\fP Jugada\&.obtenirPecaPromocionada ()"

.PP
\fBPrecondició\fP
.RS 4
- @Post S'ha retornat el tipus de peca promocionada\&. 
.RE
.PP

.SS "\fBResultatPartida\fP Jugada\&.obtenirResultatPartida ()"

.PP
\fBRetorna\fP
.RS 4
el resultat final de la partida 
.RE
.PP

.SS "Collection<\fBResultatJugada\fP> Jugada\&.obtenirResultatsJugada ()"

.PP
\fBRetorna\fP
.RS 4
\fBPosicio\fP d'origen 
.RE
.PP

.SS "\fBTauler\fP Jugada\&.obtenirTauler ()"

.PP
\fBRetorna\fP
.RS 4
\fBTauler\fP una vegada aplicada la jugada\&. 
.RE
.PP

.SS "\fBColor\fP Jugada\&.obtenirTirador ()"

.PP
\fBRetorna\fP
.RS 4
\fBColor\fP del Tirador 
.RE
.PP

.SH "Documentació de les Dades Membre"
.PP 
.SS "boolean Jugada\&.acabada\fC [private]\fP"
Indica si la partida esta acabada o no\&. 
.SS "\fBContadors\fP Jugada\&.contador\fC [private]\fP"
Cotnadors de escacs i inaccio per a cada color\&. 
.SS "\fBPosicio\fP Jugada\&.desti\fC [private]\fP"
Posició de desti de la peça\&. 
.SS "\fBColor\fP Jugada\&.guanyador\fC [private]\fP"
\fBColor\fP del guanyador de la partida 
.SS "\fBPosicio\fP Jugada\&.origen\fC [private]\fP"
Posició d'origen de la peça\&. 
.SS "\fBPeca\fP Jugada\&.pecaDesti\fC [private]\fP"
Peça del desti (posible peça del rival elimanada)\&. 
.SS "\fBPeca\fP Jugada\&.pecaOrigen\fC [private]\fP"
Peça de l'origen\&. 
.SS "\fBPeca\fP Jugada\&.pecaPromocionada\fC [private]\fP"
Peça que s'ha promocionat si escau 
.SS "\fBResultatPartida\fP Jugada\&.resultatFinal\fC [private]\fP"
Indica el resultat final de la partida\&. 
.SS "Queue<\fBResultatJugada\fP> Jugada\&.resultatsJugada\fC [private]\fP"
Llista de resultats de la jugada en questio\&. 
.SS "\fBTauler\fP Jugada\&.t\fC [private]\fP"
Guarda la classe \fBTauler\fP que s'esta utilitzant actualment\&. 
.SS "\fBColor\fP Jugada\&.tirador\fC [private]\fP"
El que realitza la jugada\&. 

.SH "Autor"
.PP 
Generat automàticament per Doxygen per a Escacs a partir del codi font\&.


src/Jugada.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedList;
import java.util.Queue;

/**
 *  @file Jugada.java
 *  @brief Classe Jugada
 */

/** 
 *  @class Jugada
 *  @brief Conte tota la informacio de una Jugada i part de les dades de una partida.
 *  @author Eloi Quintana Ferrer
 *  @author u1963141
 */

public class Jugada {
    
    /**
     * Indica si la partida esta acabada o no.
     */
    private boolean acabada;
    
    /**
     * Cotnadors de escacs i inaccio per a cada color.
     */
    private Contadors contador;
    
    /**
    * Posició de desti de la peça.
    */
    private Posicio desti;
    
    /**
    * Color del guanyador de la partida
    */
    private Color guanyador;
    
    /**
    * Posició d'origen de la peça.
    */
    private Posicio origen;
    
    /**
    * Peça del desti (posible peça del rival elimanada).
    */
    private Peca pecaDesti;
    
    /**
    * Peça de l'origen.
    */
    private Peca pecaOrigen;
    
    /**
    * Peça que s'ha promocionat si escau
    */
    private Peca pecaPromocionada;
    
    /**
     * Indica el resultat final de la partida.
     */
    private ResultatPartida resultatFinal;

    /**
    * Llista de resultats de la jugada en questio.
    */
    //private ResultatJugada resultat;
    private Queue<ResultatJugada> resultatsJugada;

    /**
     * Guarda la classe Tauler que s'esta utilitzant actualment.
     */
    private Tauler t;
    
    /**
    * El que realitza la jugada.
    */
    private Color tirador;
    
    /**
    * @pre
    * @Post constructor
    */
    public Jugada(){
    } 
    
    /**
    * @pre origen != null, desti != null, tirador != null, 
    * @pre origen i desti dins el tauler
    * @Post S'han guardat les variables d'entrada.
    * @param origen Posició d'origen de la peça.
    * @param desti Posició de desti de la peça.
    * @param tirador Tirador de la jugada.
    */
    public Jugada(Posicio origen, Posicio desti, Color tirador){
        this.origen = origen;
        this.desti = desti;
        this.tirador = tirador;
        this.resultatsJugada = new LinkedList<ResultatJugada>();
    } 
    
    /**
    * @pre origen != null, desti != null, pecaOrigen != null, 
    * @pre origen i desti dins el tauler
    * @Post S'han guardat les variables d'entrada.
    * @param origen Posició d'origen de la peça.
    * @param desti Posició de desti de la peça.
    * @param pecaOrigen Peça de l'origen.
    * @param pecaDesti Peça del desti (posible peça del rival elimanada).
    * @param resultat Resultat de la jugada.
    * @param eliminades Peces que s'han eliminat.
    */
    public Jugada(Posicio origen, Posicio desti, Peca pecaOrigen, Peca pecaDesti){
        this.origen = origen;
        this.desti = desti;
        this.pecaOrigen = pecaOrigen;
        this.pecaDesti = pecaDesti;
        this.resultatsJugada = new LinkedList<ResultatJugada>();
        this.tirador = pecaOrigen.color();
        this.t = null;
        this.acabada = false;
        this.resultatFinal = null;
        this.contador = null;
    } 
    
    /**
    * @pre resultat != null
    * @Post S'han guardat les variables d'entrada.
    * @param resultat Resultat de la Jugada.
    * @param tirador color del jugador que realitza la jugada
    */
    public Jugada(ResultatJugada resultat,Color tirador){
        this.resultatsJugada = new LinkedList<ResultatJugada>();
        afegirResultatJugada(resultat);
        this.tirador = tirador;
    } 


    /**
     * @pre res != null
     * @Post s'ha afegit a la llista el resultat de la jugada
     * @param t Tauler referenciat.
    */   
    public void afegirResultatJugada(ResultatJugada res){
        resultatsJugada.add(res);
    }
    
    /**
    * @pre t != null
    * @Post S'ha substituit la referencia de tauler amb la d'entrada
    * @param t Tauler referenciat.
    */
    public void canviTauler(Tauler t){
        this.t = t;
    }
    
    /**
     * @pre -
     * @post S'han assignat els parametrs d'entrada a les corresponents varialbes/referencies.
     * @param t Tauler referenciat.
     * @param partidaAcabada true si la partida es acabada, false altrament.
     * @param resultatFinal resultat de la partida
     * @param contador contadors de innacció i escac continu.
    */
    public void finalitzarJugada(Tauler t, boolean partidaAcabada, ResultatPartida resultatFinal, Contadors contador, Peca pecaPromocionada, Color guanyador){
        this.t = t;
        this.acabada = partidaAcabada;
        this.resultatFinal = resultatFinal;
        this.contador = contador;
        this.pecaPromocionada = pecaPromocionada;
        this.guanyador = guanyador;
    }
    
    /**
    * @return True si la partida esta acabada, false sino.
    */
    public boolean obtenirAcabada(){
        return acabada;
    }
    
    /**
    * @return Contadors de la partida.
    */    
    public Contadors obtenirContadors(){
        return contador;
    }
    
    /**
    * @return Posicio de desti
    */
    public Posicio obtenirDesti(){
        return desti;
    }
    
    /**
    * @return retorna el color del guanyador
    */
    public Color obtenirGuanyador(){
        return guanyador;
    }
    
    /**
    * @return Posicio d'origen
    */
    public Posicio obtenirOrigen(){
        return origen;
    }
    
    /**
    * @return peca de desti
    */
    public Peca obtenirPecaDesti(){
        return pecaDesti;
    }
    
    /**
    * @return peca d'origen
    */
    public Peca obtenirPecaOrigen(){
        return pecaOrigen;
    }
    
    /**
    * @pre -
    * @Post S'ha retornat el tipus de peca promocionada.
    */
    public TipusPeca obtenirPecaPromocionada(){
        if(pecaPromocionada == null){
            return null;
        }
        else{
            return pecaPromocionada.tipus();
        }
    }
    
    /**
    * @return el resultat final de la partida
    */
    public ResultatPartida obtenirResultatPartida(){
        return resultatFinal;
    }
    
    /**
    * @return Posicio d'origen
    */
    public Collection<ResultatJugada> obtenirResultatsJugada(){
        return resultatsJugada;
    }
    
    /**
    * @return Tauler una vegada aplicada la jugada.
    */
    public Tauler obtenirTauler(){
        return t;
    }
    
    /**
    * @return Color del Tirador
    */
    public Color obtenirTirador(){
        return tirador;
    }
}
